drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,931 @@
+
+	return 0;
+}
+
+static int udc_ambiq_host_wakeup(const struct device *dev)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	am_hal_usb_start_remote_wakeup(priv->usb_handle);
+
+	return 0;
+}
+
+static int udc_ambiq_set_address(const struct device *dev, const uint8_t addr)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("addr %u (0x%02x)", addr, addr);
+	am_hal_usb_set_addr(priv->usb_handle, addr);
+	am_hal_usb_set_dev_state(priv->usb_handle, AM_HAL_USB_DEV_STATE_ADDRESSED);
+
+	return 0;
+}
+
+static int udc_ambiq_test_mode(const struct device *dev, const uint8_t mode, const bool dryrun)
+{
+	am_hal_usb_test_mode_e am_usb_test_mode;
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	switch (mode) {
+	case USB_DWC2_DCTL_TSTCTL_TESTJ:
+		am_usb_test_mode = AM_HAL_USB_TEST_J;
+		break;
+	case USB_DWC2_DCTL_TSTCTL_TESTK:
+		am_usb_test_mode = AM_HAL_USB_TEST_K;
+		break;
+	case USB_DWC2_DCTL_TSTCTL_TESTSN:
+		am_usb_test_mode = AM_HAL_USB_TEST_SE0_NAK;
+		break;
+	case USB_DWC2_DCTL_TSTCTL_TESTPM:
+		am_usb_test_mode = AM_HAL_USB_TEST_PACKET;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (dryrun) {
+		LOG_DBG("Test Mode %u supported", mode);
+		return 0;
+	}
+
+	am_hal_usb_enter_test_mode(priv->usb_handle);
+	am_hal_usb_test_mode(priv->usb_handle, am_usb_test_mode);
+
+	return 0;
+}
+
+static int udc_ambiq_enable(const struct device *dev)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	/* USB soft connect */
+	am_hal_usb_attach(priv->usb_handle);
+	LOG_DBG("Enable UDC");
+
+	return 0;
+}
+
+static int udc_ambiq_disable(const struct device *dev)
+{
+	unsigned int lock_key;
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	const struct udc_ambiq_config *cfg = dev->config;
+
+	/* Disable USB interrupt */
+	lock_key = irq_lock();
+	cfg->irq_disable_func(dev);
+	irq_unlock(lock_key);
+
+	/* Disable soft disconnect */
+	am_hal_usb_detach(priv->usb_handle);
+	am_hal_usb_intr_usb_disable(priv->usb_handle, USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
+							      USB_CFG2_SuspendE_Msk |
+							      USB_CFG2_ResetE_Msk);
+	am_hal_usb_intr_usb_clear(priv->usb_handle);
+	for (unsigned int i = 0; i < cfg->num_endpoints; i++) {
+		am_hal_usb_ep_state_reset(priv->usb_handle, i);
+		am_hal_usb_ep_state_reset(priv->usb_handle, BIT(7) | i);
+	}
+	LOG_DBG("Disable UDC");
+
+	return 0;
+}
+
+static void udc_ambiq_usb_isr(const struct device *dev)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	uint32_t int_status[3];
+
+	am_hal_usb_intr_status_get(priv->usb_handle, &int_status[0], &int_status[1],
+				   &int_status[2]);
+	am_hal_usb_interrupt_service(priv->usb_handle, int_status[0], int_status[1], int_status[2]);
+}
+
+static int usb_power_rails_set(bool on)
+{
+#define USB_DEV_NODE DT_NODELABEL(zephyr_udc0)
+#if DT_NODE_HAS_STATUS(USB_DEV_NODE, okay) && DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb33_gpios) &&    \
+	DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb0p9_gpios)
```

Comment:
This is not correct. (Also, you cannot use `DT_NODELABEL(zephyr_udc0)` here).
Move vddusb33_gpio and vddusb0p9_gpio to `struct udc_ambiq_config`, use `GPIO_DT_SPEC_GET_OR(DT_DRV_INST(n), reset_gpios, {0})`;

Reply:
Thanks @jfischer-no for the pointers. Added the change to use `GPIO_DT_SPEC_GET_OR` instead.